<?xml version="1.0" encoding="utf-8" standalone="yes" ?>
<rss version="2.0" xmlns:atom="http://www.w3.org/2005/Atom">
  <channel>
    <title>Openstack on hablog</title>
    <link>http://blog.osamu.habuka.jp/tags/openstack/</link>
    <description>Recent content in Openstack on hablog</description>
    <generator>Hugo -- gohugo.io</generator>
    <language>ja-jp</language>
    <lastBuildDate>Thu, 04 Dec 2014 08:08:08 +0900</lastBuildDate>
    <atom:link href="http://blog.osamu.habuka.jp/tags/openstack/index.xml" rel="self" type="application/rss+xml" />
    
    <item>
      <title>OpenStack Advent Calendar 2014 4日目</title>
      <link>http://blog.osamu.habuka.jp/post/2014-12-03-openstack-advent-calendar-2014-4th/</link>
      <pubDate>Thu, 04 Dec 2014 08:08:08 +0900</pubDate>
      
      <guid>http://blog.osamu.habuka.jp/post/2014-12-03-openstack-advent-calendar-2014-4th/</guid>
      <description>

&lt;p&gt;これは &lt;a href=&#34;http://www.adventar.org/calendars/569&#34;&gt;OpenStack Advent Calendar 2014&lt;/a&gt; の4日目のエントリです。昨日は日本 Eucalyptus ユーザ会のエントリ &lt;a href=&#34;http://eucalyptus-users.jp/blog/2014/12/03/openstack-on-eucalyptus-slash-openstack-advent-calendar-2014-dot-12-dot-03/&#34;&gt;OpenStack on Eucalyptus&lt;/a&gt; でした。&lt;/p&gt;

&lt;p&gt;今日は題して「Eucalyptus ユーザが解説する Nova API 改造指南」です。例年ですと私の OpenStack ネタのエントリは失敗するのがお約束ですが、今日も無事失敗できるかどうかドキドキです。&lt;/p&gt;

&lt;p&gt;で、このエントリの目標は「API にパラメータを何か追加してみて、ちゃんとそれが処理されるか否か」という緩い目標設定となっております。(あまりマトモな目標設定をすると失敗する以前にまず何もできなかったりするので…」&lt;/p&gt;

&lt;h2 id=&#34;出力される項目を増やす&#34;&gt;出力される項目を増やす&lt;/h2&gt;

&lt;p&gt;まぁ、誰もがやってみる改造ですね。例えば nova list の出力にはそのインスタンスが「どの nova-compute で起動しているか？」なんて情報は含まれていません。知りたい場合は DB を覗かないと知れません(ホント？)、これはとても面倒臭い…ということで、nova list の結果にその情報が含まれるようにしちゃいましょう。&lt;/p&gt;

&lt;pre&gt;&lt;code&gt;$ nova list
+--------------------------------------+------+--------+------------+-------------+------------------+
| ID                                   | Name | Status | Task State | Power State | Networks         |
+--------------------------------------+------+--------+------------+-------------+------------------+
| 98f7c56e-310f-4e57-9c9a-c0c5aa48cf08 | vm01 | ACTIVE | -          | Running     | private=10.0.0.2 |
+--------------------------------------+------+--------+------------+-------------+------------------+
&lt;/code&gt;&lt;/pre&gt;

&lt;p&gt;まず、DB「nova」のテーブル「instances」を確認すると、以下のフィールドがあります。&lt;/p&gt;

&lt;pre&gt;&lt;code&gt;+--------------------------+-----------------------+------+-----+---------+----------------+
| Field                    | Type                  | Null | Key | Default | Extra          |
+--------------------------+-----------------------+------+-----+---------+----------------+
| created_at               | datetime              | YES  |     | NULL    |                |
| updated_at               | datetime              | YES  |     | NULL    |                |
| deleted_at               | datetime              | YES  |     | NULL    |                |
| id                       | int(11)               | NO   | PRI | NULL    | auto_increment |
| internal_id              | int(11)               | YES  |     | NULL    |                |
| user_id                  | varchar(255)          | YES  |     | NULL    |                |
| project_id               | varchar(255)          | YES  | MUL | NULL    |                |
| image_ref                | varchar(255)          | YES  |     | NULL    |                |
| kernel_id                | varchar(255)          | YES  |     | NULL    |                |
| ramdisk_id               | varchar(255)          | YES  |     | NULL    |                |
| launch_index             | int(11)               | YES  |     | NULL    |                |
| key_name                 | varchar(255)          | YES  |     | NULL    |                |
| key_data                 | mediumtext            | YES  |     | NULL    |                |
| power_state              | int(11)               | YES  |     | NULL    |                |
| vm_state                 | varchar(255)          | YES  |     | NULL    |                |
| memory_mb                | int(11)               | YES  |     | NULL    |                |
| vcpus                    | int(11)               | YES  |     | NULL    |                |
| hostname                 | varchar(255)          | YES  |     | NULL    |                |
| host                     | varchar(255)          | YES  | MUL | NULL    |                |
| user_data                | mediumtext            | YES  |     | NULL    |                |
| reservation_id           | varchar(255)          | YES  | MUL | NULL    |                |
| scheduled_at             | datetime              | YES  |     | NULL    |                |
| launched_at              | datetime              | YES  |     | NULL    |                |
| terminated_at            | datetime              | YES  | MUL | NULL    |                |
| display_name             | varchar(255)          | YES  |     | NULL    |                |
| display_description      | varchar(255)          | YES  |     | NULL    |                |
| availability_zone        | varchar(255)          | YES  |     | NULL    |                |
| locked                   | tinyint(1)            | YES  |     | NULL    |                |
| os_type                  | varchar(255)          | YES  |     | NULL    |                |
| launched_on              | mediumtext            | YES  |     | NULL    |                |
| instance_type_id         | int(11)               | YES  |     | NULL    |                |
| vm_mode                  | varchar(255)          | YES  |     | NULL    |                |
| uuid                     | varchar(36)           | YES  | UNI | NULL    |                |
| architecture             | varchar(255)          | YES  |     | NULL    |                |
| root_device_name         | varchar(255)          | YES  |     | NULL    |                |
| access_ip_v4             | varchar(39)           | YES  |     | NULL    |                |
| access_ip_v6             | varchar(39)           | YES  |     | NULL    |                |
| config_drive             | varchar(255)          | YES  |     | NULL    |                |
| task_state               | varchar(255)          | YES  | MUL | NULL    |                |
| default_ephemeral_device | varchar(255)          | YES  |     | NULL    |                |
| default_swap_device      | varchar(255)          | YES  |     | NULL    |                |
| progress                 | int(11)               | YES  |     | NULL    |                |
| auto_disk_config         | tinyint(1)            | YES  |     | NULL    |                |
| shutdown_terminate       | tinyint(1)            | YES  |     | NULL    |                |
| disable_terminate        | tinyint(1)            | YES  |     | NULL    |                |
| root_gb                  | int(11)               | YES  |     | NULL    |                |
| ephemeral_gb             | int(11)               | YES  |     | NULL    |                |
| cell_name                | varchar(255)          | YES  |     | NULL    |                |
| node                     | varchar(255)          | YES  |     | NULL    |                |
| deleted                  | int(11)               | YES  |     | NULL    |                |
| locked_by                | enum(&#39;owner&#39;,&#39;admin&#39;) | YES  |     | NULL    |                |
| cleaned                  | int(11)               | YES  |     | NULL    |                |
| ephemeral_key_uuid       | varchar(36)           | YES  |     | NULL    |                |
+--------------------------+-----------------------+------+-----+---------+----------------+
&lt;/code&gt;&lt;/pre&gt;

&lt;p&gt;で、これらのフィールドのなかで、host と node がそれっぽいのですが、それらのどちらが nova-compute のホスト名なのかは、にわかユーザにはわからないので、どちらも表示させることにします。&lt;/p&gt;

&lt;p&gt;……&lt;/p&gt;

&lt;p&gt;えっと…まるで初心者のような──いや、僕は確かに初心者なので間違ってはいないのですが、不思議なことが起きました。何かと言うと、/opt/stack/nova が消えました。とりあえずコマンド履歴を途中から確認してみました。&lt;/p&gt;

&lt;pre&gt;&lt;code&gt;ubuntu@ubuntu:/opt/stack/nova$ vim nova/db/sqlalchemy/api.py
ubuntu@ubuntu:/opt/stack/nova$ vim nova/api/openstack/compute/servers.py
ubuntu@ubuntu:/opt/stack/nova$ vim nova/compute/api.py
ubuntu@ubuntu:/opt/stack/nova$ vim /usr/local/lib/python2.7/dist-packages/py[TAB]
ubuntu@ubuntu:/opt/stack/nova$ vim /usr/local/lib/python2.7/dist-packages/py[TAB]
ubuntu@ubuntu:/opt/stack/nova$ vim /usr/local/lib/python2.7/dist-packages/py[TAB]
ubuntu@ubuntu:/opt/stack/nova$ vim /usr/local/lib/python2.7/dist-packages/novaclient/[TAB]
ubuntu@ubuntu:/opt/stack/nova$ vim /usr/local/lib/python2.7/dist-packages/novaclient/v1_1/[TAB]
ubuntu@ubuntu:/opt/stack/nova$ vim /usr/local/lib/python2.7/dist-packages/novaclient/v1_1/[TAB]
ubuntu@ubuntu:/opt/stack/nova$ vim /usr/local/lib/python2.7/dist-packages/novaclient/v1_1/[TAB]
ubuntu@ubuntu:/opt/stack/nova$ vim /usr/local/lib/python2.7/dist-packages/novaclient/v1_1/[TAB]
ubuntu@ubuntu:/opt/stack/nova$ vim /usr/local/lib/python2.7/dist-packages/novaclient/v1_1/se[TAB]
ubuntu@ubuntu:/opt/stack/nova$ vim /usr/local/lib/python2.7/dist-packages/novaclient/v1_1/serv[TAB]
ubuntu@ubuntu:/opt/stack/nova$ vim /usr/local/lib/python2.7/dist-packages/novaclient/v1_1/server[TAB]
ubuntu@ubuntu:/opt/stack/nova$ vim /usr/local/lib/python2.7/dist-packages/novaclient/v1_1/servers.py
ubuntu@ubuntu:/opt/stack/nova$ vim /usr/local/lib/python2.7/dist-packages/novaclient/v1_1/servers.py
ubuntu@ubuntu:/opt/stack/nova$ vim nova/db/sqlalchemy/api.py
ubuntu@ubuntu:/opt/stack/nova$ vim nova/db/sqlalchemy/api.py
ubuntu@ubuntu:/opt/stack/nova$ pwd
/opt/stack/nova
ubuntu@ubuntu:/opt/stack/nova$ ls -al
total 0
ubuntu@ubuntu:/opt/stack/nova$ df -h
Filesystem      Size  Used Avail Use% Mounted on
/dev/vda1        16G  4.8G   11G  33% /
none            4.0K     0  4.0K   0% /sys/fs/cgroup
udev            2.0G  4.0K  2.0G   1% /dev
tmpfs           396M  432K  395M   1% /run
none            5.0M     0  5.0M   0% /run/lock
none            2.0G     0  2.0G   0% /run/shm
none            100M     0  100M   0% /run/user
ubuntu@ubuntu:/opt/stack/nova$ ls -al /opt/
total 12
drwxr-xr-x  3 root   root 4096 Dec  4 15:10 .
drwxr-xr-x 22 root   root 4096 Jul  6 11:45 ..
drwxr-xr-x 14 ubuntu root 4096 Dec  4 19:10 stack
ubuntu@ubuntu:/opt/stack/nova$ ls -al /opt/stack/
total 56
drwxr-xr-x 14 ubuntu root   4096 Dec  4 19:10 .
drwxr-xr-x  3 root   root   4096 Dec  4 15:10 ..
drwxr-xr-x 10 ubuntu ubuntu 4096 Dec  4 17:18 cinder
drwxr-xr-x  6 ubuntu root   4096 Dec  4 17:32 data
drwxr-xr-x  9 ubuntu ubuntu 4096 Dec  4 17:18 glance
drwxr-xr-x 12 ubuntu ubuntu 4096 Dec  4 17:23 heat
drwxr-xr-x  7 ubuntu ubuntu 4096 Dec  4 17:22 heat-cfntools
drwxr-xr-x 10 ubuntu ubuntu 4096 Dec  4 17:22 heat-templates
drwxr-xr-x 11 ubuntu ubuntu 4096 Dec  4 17:24 horizon
drwxr-xr-x 12 ubuntu ubuntu 4096 Dec  4 17:18 keystone
drwxr-xr-x  9 ubuntu ubuntu 4096 Dec  4 16:40 noVNC
drwxr-xr-x  5 ubuntu ubuntu 4096 Dec  4 16:16 requirements
drwxr-xr-x  3 ubuntu ubuntu 4096 Dec  4 17:23 status
drwxr-xr-x  8 ubuntu ubuntu 4096 Dec  4 17:32 tempest
&lt;/code&gt;&lt;/pre&gt;

&lt;p&gt;いやぁ〜、いくら僕が OpenStack 初心者だからと言って、こんな現象に遭遇するとは…。一体どんな操作ミスをしたらこんなことになるんでしょう…不思議です。&lt;/p&gt;

&lt;p&gt;まぁ、でも消えちゃったものは仕方ないので、./unstack.sh 叩いてもう一度 ./stack.sh を叩くことにします。&lt;/p&gt;

&lt;p&gt;で、気をとりなおして再挑戦…と思って調査したんですが、&lt;a href=&#34;https://github.com/openstack/nova/blob/master/nova/api/openstack/compute/servers.py#L604-L610&#34;&gt;api/openstack/compute/servers.py のここ&lt;/a&gt; でインスタンスのリストを取得している風というのは分かるんですが、この結果を nova list 用に整形している &lt;a href=&#34;https://github.com/openstack/nova/blob/master/nova/api/openstack/compute/servers.py#L622&#34;&gt;この部分&lt;/a&gt; から先の&lt;a href=&#34;https://github.com/openstack/nova/blob/master/nova/api/openstack/compute/views/servers.py#L117-L120&#34;&gt;ここあたり&lt;/a&gt;で追うのに行き詰まりました。&lt;/p&gt;

&lt;p&gt;ここら辺が非 OpenStacker の限界ということでしょうか…orz&lt;/p&gt;

&lt;p&gt;ちなみに nova list ではなく、nova baremetal-node-list の出力項目を変更したい場合は &lt;a href=&#34;https://github.com/openstack/nova/blob/master/nova/api/openstack/compute/contrib/baremetal_nodes.py#L32-L33&#34;&gt;この node_fields の要素をイジって&lt;/a&gt;、更に novaclient の &lt;a href=&#34;https://github.com/openstack/python-novaclient/blob/master/novaclient/v1_1/contrib/baremetal.py#L247-L259&#34;&gt;v1_1/contrib/baremetal.py の _print_baremetal_nodes_list のここ&lt;/a&gt;をイジれば変更できます…が、たぶんそんなこと OpenStacker の皆さんなら周知の事実ですよね…すみません…orz&lt;/p&gt;

&lt;hr /&gt;

&lt;p&gt;そのあと元木さんから「それ、改造しなくても admin なら &amp;ndash;fields オプションで OS-EXT-SRV-ATTR: で引けるよ？」と教えてもらいましたwwww (同じく真壁さんからもw)&lt;/p&gt;

&lt;pre&gt;&lt;code&gt;$ nova list --all-tenants --fields name,networks,OS-EXT-SRV-ATTR:host,OS-EXT-SRV-ATTR:node
+--------------------------------------+------+------------------+-----------------------+-----------------------+
| ID                                   | Name | Networks         | OS-EXT-SRV-ATTR: Host | OS-EXT-SRV-ATTR: Node |
+--------------------------------------+------+------------------+-----------------------+-----------------------+
| 9bc0c1fb-d55c-481f-84fb-fadd21ec1d35 | vm01 | private=10.0.0.2 | ubuntu                |                       |
+--------------------------------------+------+------------------+-----------------------+-----------------------+
&lt;/code&gt;&lt;/pre&gt;

&lt;h2 id=&#34;novaclient-のサブコマンドに引数を持たないオプションを追加する方法&#34;&gt;novaclient のサブコマンドに引数を持たないオプションを追加する方法&lt;/h2&gt;

&lt;p&gt;例えば、nova baremetal-node-delete のオプションに引数を持たない &amp;ndash;force というオプションを追加したい場合は &lt;a href=&#34;https://github.com/openstack/python-novaclient/blob/master/novaclient/v1_1/contrib/baremetal.py&#34;&gt;v1_1/contrib/baremetal.py&lt;/a&gt; の以下を&lt;/p&gt;

&lt;pre&gt;&lt;code class=&#34;language-python&#34;&gt;@utils.arg(
    &#39;node&#39;,
    metavar=&#39;&amp;lt;node&amp;gt;&#39;,
    help=_(&#39;ID of the node to delete.&#39;))
def do_baremetal_node_delete(cs, args):
    &amp;quot;&amp;quot;&amp;quot;Remove a baremetal node and any associated interfaces.&amp;quot;&amp;quot;&amp;quot;
    node = _find_baremetal_node(cs, args.node)
    cs.baremetal.delete(node)
&lt;/code&gt;&lt;/pre&gt;

&lt;p&gt;以下のように変更するだけです。&lt;/p&gt;

&lt;pre&gt;&lt;code class=&#34;language-python&#34;&gt;@utils.arg(
    &#39;node&#39;,
    metavar=&#39;&amp;lt;node&amp;gt;&#39;,
    help=_(&#39;ID of the node to delete.&#39;))
@utils.arg(&#39;--force&#39;,
    action=&amp;quot;store_true&amp;quot;,
    default=False,
    help=&#39;force delete baremetal machine.&#39;)
def do_baremetal_node_delete(cs, args):
    &amp;quot;&amp;quot;&amp;quot;Remove a baremetal node and any associated interfaces.&amp;quot;&amp;quot;&amp;quot;
    node = _find_baremetal_node(cs, args.node)
    cs.baremetal.delete(node)
&lt;/code&gt;&lt;/pre&gt;

&lt;p&gt;あとは do_baremetal_node_delete の中でオプションの値を参照する場合は args.force を見れば OK です。&lt;/p&gt;

&lt;hr /&gt;

&lt;p&gt;明日の &lt;a href=&#34;http://www.adventar.org/calendars/569&#34;&gt;OpenStack Advent Calendar 2014 5日目&lt;/a&gt; は Shogos3 さんのエントリです。&lt;/p&gt;
</description>
    </item>
    
    <item>
      <title>OpenStack Advent Calendar 2013 JP 8日目</title>
      <link>http://blog.osamu.habuka.jp/post/2013-12-08.AdventCalendar-day-8/</link>
      <pubDate>Sun, 08 Dec 2013 01:11:57 +0900</pubDate>
      
      <guid>http://blog.osamu.habuka.jp/post/2013-12-08.AdventCalendar-day-8/</guid>
      <description>

&lt;h1 id=&#34;碧川涼&#34;&gt;碧川涼&lt;/h1&gt;

&lt;p&gt;さて、今日も名前に緑色のつく人を見つけられませんでした。が、よくよく考えたら「碧」っていう字も、青だったり緑だったりするじゃないですか、ってことで、今日はゲーム『NOëL 〜La neige〜』に登場するキャラクターの「&lt;a href=&#34;http://ja.wikipedia.org/wiki/NOeL#NO.C3.ABL_.E3.80.9CLa_neige.E3.80.9C_.2F_NO.C3.ABL_.E3.80.9CLa_neige.E3.80.9C_Special&#34;&gt;碧川涼&lt;/a&gt;」さんの誕生日だそうです。お誕生日おめでとうございます。えーと、恋愛シミュレーションゲームだそうです。すみません、残念ながらやったことないので、存じあげておりませんでした。&lt;/p&gt;

&lt;p&gt;それと、この碧川涼というキャラクターの声を演じているのは&lt;a href=&#34;http://ja.wikipedia.org/wiki/%E6%A0%B9%E8%B0%B7%E7%BE%8E%E6%99%BA%E5%AD%90&#34;&gt;根谷美智子&lt;/a&gt;さんなんですが、根谷さんの親友の&lt;a href=&#34;http://ja.wikipedia.org/wiki/%E8%8F%85%E5%8E%9F%E7%A5%A5%E5%AD%90&#34;&gt;菅原祥子&lt;/a&gt;さんは12月12日が誕生日だそうです。おめでとうございます。&lt;/p&gt;

&lt;h1 id=&#34;glance-のかわりに-walrus-を使ってみる&#34;&gt;Glance のかわりに Walrus を使ってみる&lt;/h1&gt;

&lt;p&gt;まず最初に、&lt;a href=&#34;https://twitter.com/ishikawa84g&#34;&gt;我らが愛する石川&lt;/a&gt; さんがこんなことをツブやいてまして&lt;/p&gt;

&lt;p&gt;&lt;blockquote class=&#34;twitter-tweet&#34; data-lang=&#34;ja&#34;&gt;&lt;p lang=&#34;ja&#34; dir=&#34;ltr&#34;&gt;OpenStack Advent Calendar 2013、誰も記事内で AdventCalendarの何日目という宣言をしてないの良くない。&lt;/p&gt;&amp;mdash; 石川@驚きの白々しさ (@ishikawa84g) &lt;a href=&#34;https://twitter.com/ishikawa84g/status/408833430890115072&#34;&gt;2013年12月6日&lt;/a&gt;&lt;/blockquote&gt;
&lt;script async src=&#34;//platform.twitter.com/widgets.js&#34; charset=&#34;utf-8&#34;&gt;&lt;/script&gt;&lt;/p&gt;

&lt;p&gt;これは、やばい、このままだと enforcing されちゃう…(((( ；ﾟДﾟ)))…ってことで、今日は OpenStack Advent Calendar 2013 JP の 8 日目です。&lt;/p&gt;

&lt;p&gt;ところで、&lt;a href=&#34;http://036.habuka.jp/diary/?date=20121203#p04&#34;&gt;去年の OpenStack Advent Calendar 2012 JP の僕のエントリー&lt;/a&gt;は惨々な結果でした…。今年は反省しちゃんとやりきりたいと思いますが、どうやら僕は OpenStack 界隈では「負け犬」「噛ませ犬」のポジションらしく「今年も&lt;font size=&#34;5&#34;&gt;惨々な内容&lt;/font&gt;でひとつよろしく頼むよキミぃ」とか言われてます…(｡´Д⊂)&lt;/p&gt;

&lt;h2 id=&#34;openstack-環境の設定&#34;&gt;OpenStack 環境の設定&lt;/h2&gt;

&lt;p&gt;えー、僕は OpenStack 初心者&lt;sup class=&#34;footnote-ref&#34; id=&#34;fnref:1&#34;&gt;&lt;a rel=&#34;footnote&#34; href=&#34;#fn:1&#34;&gt;1&lt;/a&gt;&lt;/sup&gt;なので、いま流行りの &lt;a href=&#34;http://openstack.redhat.com/Quickstart&#34;&gt;RDO? Packstack?&lt;/a&gt; を使って構築しました。まぁ、Packstack を用いた構築方法についてはきっと誰かが今回の Advent Calendar で書いてくれるでしょうから、ここでは割愛します。&lt;/p&gt;

&lt;p&gt;ちなみに、ここで使う Walrus は先日の &lt;a href=&#34;http://www.adventar.org/calendars/202&#34;&gt;Eucalyptus Advent Calendar 2013&lt;/a&gt; のエントリ「&lt;a href=&#34;http://036.habuka.jp/diary/?date=20131202&#34;&gt;Walrus をだけを使ってみる&lt;/a&gt;」で作った環境を使います。&lt;/p&gt;

&lt;p&gt;まず、glance のサービスを止めます。だって使わないんだもんだもん。止めちゃいますよー。&lt;/p&gt;

&lt;pre&gt;&lt;code&gt;[root@rdo ~(keystone_admin)]# /etc/init.d/openstack-glance-api stop
Stopping openstack-glance-api:                             [  OK  ]
[root@rdo ~(keystone_admin)]# /etc/init.d/openstack-glance-registry stop
Stopping openstack-glance-registry:                        [  OK  ]
[root@rdo ~(keystone_admin)]# /etc/init.d/openstack-glance-scrubber stop
&lt;/code&gt;&lt;/pre&gt;

&lt;p&gt;で、nova が glance じゃなくて Walrus を使うように設定します。&lt;/p&gt;

&lt;pre&gt;&lt;code&gt;[root@rdo ~(keystone_admin)]# diff -u /etc/nova/nova.conf.glance /etc/nova/nova.conf
--- /etc/nova/nova.conf.glance  2013-12-03 14:46:31.314000061 +0900
+++ /etc/nova/nova.conf 2013-12-03 17:29:31.213000032 +0900
@@ -939,7 +939,7 @@
 # with https:// for ssl-based glance api servers.
 # ([hostname|ip]:port) (list value)
 #glance_api_servers=$glance_host:$glance_port
-glance_api_servers=192.168.32.74:9292
+#glance_api_servers=192.168.32.74:9292

 # Allow to perform insecure SSL (https) requests to glance
 # (boolean value)
@@ -965,16 +965,16 @@

 # hostname or ip for OpenStack to use when accessing the s3
 # api (string value)
-#s3_host=$my_ip
+s3_host=192.168.32.73

 # port used when accessing the s3 api (integer value)
-#s3_port=3333
+s3_port=8773

 # access key to use for s3 server for images (string value)
-#s3_access_key=notchecked
+s3_access_key=AKIAMNYEAUG5ZJBN9KPH

 # secret key to use for s3 server for images (string value)
-#s3_secret_key=notchecked
+s3_secret_key=WlsMz8d6LdZKOmSQo07LmtujRqN6OVKw4NoKe02V

 # whether to use ssl when talking to s3 (boolean value)
 #s3_use_ssl=false
@@ -2595,7 +2595,7 @@
 qpid_reconnect=True
 sql_connection=mysql://nova:db56f39a7e064ccc@192.168.32.74/nova
 qpid_reconnect_timeout=0
-image_service=nova.image.glance.GlanceImageService
+image_service=nova.image.s3.S3ImageService
 logdir=/var/log/nova
 qpid_reconnect_interval_max=0
 qpid_reconnect_limit=0
&lt;/code&gt;&lt;/pre&gt;

&lt;p&gt;えーと、ここまで設定して、ふと気付いたんですよ。Walrus のエンドポイントって http://ホスト:8773/services/Walrus っていう形式なんすよねー。えぇ、その値を設定するパラメータって s3_host と s3_port じゃ書ききれないよねぇ…。&lt;/p&gt;

&lt;p&gt;で、nova/image/s3.py を書き換えようと思ったんですが、@ishikawa84g さんと @shida1234 さんが「それ、keystone に登録しちゃえばいいんじゃない？」と助け船を出してくれまして、以下のように登録してみました。あ、もちろん keystone service-delete で glance のサービスとエンドポイント情報を削除してから実施してます。&lt;/p&gt;

&lt;pre&gt;&lt;code&gt;[root@rdo ~(keystone_admin)]# keystone service-create --name walrus --type image --description &amp;quot;Eucalyptus Image Service&amp;quot;
+-------------+----------------------------------+
|   Property  |              Value               |
+-------------+----------------------------------+
| description |     Eucalyptus Image Service     |
|      id     | af7ba823c7ab4f58917af6a0fad81006 |
|     name    |              walrus              |
|     type    |              image               |
+-------------+----------------------------------+

[root@rdo ~(keystone_admin)]# keystone endpoint-create --service af7ba823c7ab4f58917af6a0fad81006 --publicurl http://192.168.32.73:8773/services/Walrus --adminurl http://192.168.32.73:8773/services/Walrus --internalurl http://192.168.32.73:8773/services/Walrus
+-------------+-------------------------------------------+
|   Property  |                   Value                   |
+-------------+-------------------------------------------+
|   adminurl  | http://192.168.32.73:8773/services/Walrus |
|      id     |      fa3de0db20ae4903a429c35beabeed37     |
| internalurl | http://192.168.32.73:8773/services/Walrus |
|  publicurl  | http://192.168.32.73:8773/services/Walrus |
|    region   |                 regionOne                 |
|  service_id |      af7ba823c7ab4f58917af6a0fad81006     |
+-------------+-------------------------------------------+
&lt;/code&gt;&lt;/pre&gt;

&lt;p&gt;あ、一応 glance っていうか、Image Store とおしゃべりするであろう nova-api と nova-compute は再起動させます。&lt;/p&gt;

&lt;pre&gt;&lt;code&gt;[root@rdo ~(keystone_admin)]# /etc/init.d/openstack-nova-api restart
Stopping openstack-nova-api:                               [  OK  ]
Starting openstack-nova-api:                               [  OK  ]
[root@rdo ~(keystone_admin)]# /etc/init.d/openstack-nova-compute restart
Stopping openstack-nova-compute:                           [  OK  ]
Starting openstack-nova-compute:                           [  OK  ]
&lt;/code&gt;&lt;/pre&gt;

&lt;p&gt;で、ちゃんと接続できるかなーっと nova image-list (カチャカチャカチャ…ッターン！)&lt;/p&gt;

&lt;pre&gt;&lt;code&gt;[root@rdo ~(keystone_admin)]# nova image-list
ERROR: The server has either erred or is incapable of performing the requested operation. (HTTP 500) (Request-ID: req-ca8b1ed6-7ff0-4863-a05e-f88598a31469)
&lt;/code&gt;&lt;/pre&gt;

&lt;p&gt;あ、あれー？何だろう？と思って nova-api のログを見ると…&lt;/p&gt;

&lt;pre&gt;&lt;code&gt;2013-12-03 18:30:05.838 25016 TRACE nova.api.openstack   File &amp;quot;/usr/lib/python2.6/site-packages/nova/api/openstack/compute/images.py&amp;quot;, line 203, in detail
2013-12-03 18:30:05.838 25016 TRACE nova.api.openstack     **page_params)
2013-12-03 18:30:05.838 25016 TRACE nova.api.openstack   File &amp;quot;/usr/lib/python2.6/site-packages/nova/image/glance.py&amp;quot;, line 264, in detail
2013-12-03 18:30:05.838 25016 TRACE nova.api.openstack     for image in images:
2013-12-03 18:30:05.838 25016 TRACE nova.api.openstack   File &amp;quot;/usr/lib/python2.6/site-packages/glanceclient/v1/images.py&amp;quot;, line 174, in paginate
2013-12-03 18:30:05.838 25016 TRACE nova.api.openstack     images = self._list(url, &amp;quot;images&amp;quot;)
2013-12-03 18:30:05.838 25016 TRACE nova.api.openstack   File &amp;quot;/usr/lib/python2.6/site-packages/glanceclient/common/base.py&amp;quot;, line 53, in _list
2013-12-03 18:30:05.838 25016 TRACE nova.api.openstack     resp, body = self.api.json_request(&#39;GET&#39;, url)
2013-12-03 18:30:05.838 25016 TRACE nova.api.openstack   File &amp;quot;/usr/lib/python2.6/site-packages/glanceclient/common/http.py&amp;quot;, line 266, in json_request
2013-12-03 18:30:05.838 25016 TRACE nova.api.openstack     resp, body_iter = self._http_request(url, method, **kwargs)
2013-12-03 18:30:05.838 25016 TRACE nova.api.openstack   File &amp;quot;/usr/lib/python2.6/site-packages/glanceclient/common/http.py&amp;quot;, line 235, in _http_request
2013-12-03 18:30:05.838 25016 TRACE nova.api.openstack     raise exc.CommunicationError(message=message)
2013-12-03 18:30:05.838 25016 TRACE nova.api.openstack CommunicationError: Error communicating with http://192.168.32.74:9292 [Errno 111] ECONNREFUSED
&lt;/code&gt;&lt;/pre&gt;

&lt;p&gt;あらやだ、この子、Glance 見に行ってんじゃないの…。っていうか、何で nova/image/glance.py が呼ばれてるのさ…。ということでコード nova/api/openstack/compute/images.py を確認すると…&lt;/p&gt;

&lt;pre&gt;&lt;code&gt;import nova.image.glance
from nova.openstack.common.gettextutils import _
import nova.utils
&lt;/code&gt;&lt;/pre&gt;

&lt;p&gt;あぁ…、思いっきり import nova.image.glance してるじゃん…、じゃぁ誰が「image_service=nova.image.s3.S3ImageService」の設定を見てくれるの？&lt;/p&gt;

&lt;pre&gt;&lt;code&gt;[root@rdo ~(keystone_admin)]# grep -i CONF.image_service -r /usr/lib/python2.6/site-packages/nova/
[root@rdo ~(keystone_admin)]# grep &amp;quot;&#39;image_service&#39;&amp;quot; -r /usr/lib/python2.6/site-packages/nova/
[root@rdo ~(keystone_admin)]#
&lt;/code&gt;&lt;/pre&gt;

&lt;p&gt;ちょっと待て、本来なら今頃、OpenStack から Walrus を使ってインスタンスを起動して&lt;font size=&#34;5&#34;&gt;感無量～人´ω｀*)&lt;/font&gt;とかやってるはずなのに、何？あれ？誰も nova.conf の image_service を見てないんじゃないの？ちょっと！どーゆーこと？何この&lt;font size=&#34;5&#34;&gt;無力感…(:.;´;Д;ﾟ;.:)&lt;/font&gt;&lt;/p&gt;

&lt;p&gt;Advent Calendar のエントリに「Glance のかわりに Walrus を使う(キリッ」とか書いたのに、このままではまた去年と同様に&lt;font size=&#34;5&#34;&gt;「m9(^Д^)プギャー」&lt;/font&gt;ですよ。もうね、どんだけ僕は成功しちゃいけない星のもとに生まれたんすか？&lt;/p&gt;

&lt;p&gt;ってことで、足掻いてみます。そりゃーもう必死に足掻いてみます。アラフォー男児が本気で足掻いたら、きっと nova だって「おっぉ、わかったぉ、Glance やめて S3 喋ってみるよ」って折れてくれるはず、nova 懐深い子、きっと僕も包んでくれるはず。&lt;/p&gt;

&lt;p&gt;まず、手始めに nova/image/glance.py を呼び出している nova/api/openstack/compute/images.py を見てみます。&lt;/p&gt;

&lt;pre&gt;&lt;code&gt;class Controller(wsgi.Controller):
    &amp;quot;&amp;quot;&amp;quot;Base controller for retrieving/displaying images.&amp;quot;&amp;quot;&amp;quot;

    _view_builder_class = views_images.ViewBuilder

    def __init__(self, image_service=None, **kwargs):
        &amp;quot;&amp;quot;&amp;quot;Initialize new `ImageController`.

        :param image_service: `nova.image.glance:GlanceImageService`

        &amp;quot;&amp;quot;&amp;quot;
        super(Controller, self).__init__(**kwargs)
        self._image_service = (image_service or
                               nova.image.glance.get_default_image_service())
&lt;/code&gt;&lt;/pre&gt;

&lt;p&gt;うん、あら？、ほう、image_service image_service が None じゃなければ glance.get_default_image_service は呼ばれないのか。そうかそうか、んじゃー、glance.py に get_default_image_service があるってことは、s3.py にもあるのかな？&lt;/p&gt;

&lt;pre&gt;&lt;code&gt;[root@rdo ~]# grep &amp;quot;def get_default_image_service&amp;quot; -r /usr/lib/python2.6/site-packages/nova/image/s3.py | wc -l
0
&lt;/code&gt;&lt;/pre&gt;

&lt;p&gt;うん、無いね、全く無いね、容赦なく無いね。&lt;/p&gt;

&lt;p&gt;よし、おいちゃん覚悟した、Pythonian じゃないうえに Python 書けないけど、頑張る。nova/api/openstack/compute/images.py を書き換えるよ。&lt;/p&gt;

&lt;pre&gt;&lt;code&gt;--- /usr/lib/python2.6/site-packages/nova/api/openstack/compute/images.py.orig  2013-12-07 05:19:23.774000083 +0900
+++ /usr/lib/python2.6/site-packages/nova/api/openstack/compute/images.py       2013-12-07 05:14:58.438000084 +0900
@@ -21,6 +21,7 @@
 from nova.api.openstack import xmlutil
 from nova import exception
 import nova.image.glance
+import nova.image.s3
 from nova.openstack.common.gettextutils import _
 import nova.utils

@@ -96,6 +97,7 @@

         &amp;quot;&amp;quot;&amp;quot;
         super(Controller, self).__init__(**kwargs)
+        image_service = nova.image.s3.S3ImageService()
         self._image_service = (image_service or
                                nova.image.glance.get_default_image_service())

&lt;/code&gt;&lt;/pre&gt;

&lt;p&gt;で、nova-api を再起動して、nova image-list (カチャカチャカチャ…ッターン！)&lt;/p&gt;

&lt;pre&gt;&lt;code&gt;2013-12-07 05:15:24.699 10379 TRACE nova.api.openstack   File &amp;quot;/usr/lib/python2.6/site-packages/nova/api/openstack/compute/images.py&amp;quot;, line 205, in detail
2013-12-07 05:15:24.699 10379 TRACE nova.api.openstack     **page_params)
2013-12-07 05:15:24.699 10379 TRACE nova.api.openstack   File &amp;quot;/usr/lib/python2.6/site-packages/nova/image/s3.py&amp;quot;, line 168, in detail
2013-12-07 05:15:24.699 10379 TRACE nova.api.openstack     images = self.service.detail(context, **kwargs)
2013-12-07 05:15:24.699 10379 TRACE nova.api.openstack   File &amp;quot;/usr/lib/python2.6/site-packages/nova/image/glance.py&amp;quot;, line 264, in detail
2013-12-07 05:15:24.699 10379 TRACE nova.api.openstack     for image in images:
(中略)
2013-12-07 05:15:24.699 10379 TRACE nova.api.openstack   File &amp;quot;/usr/lib/python2.6/site-packages/glanceclient/common/http.py&amp;quot;, line 235, in _http_request
2013-12-07 05:15:24.699 10379 TRACE nova.api.openstack     raise exc.CommunicationError(message=message)
2013-12-07 05:15:24.699 10379 TRACE nova.api.openstack CommunicationError: Error communicating with http://192.168.32.74:9292 [Errno 111] ECONNREFUSED
&lt;/code&gt;&lt;/pre&gt;

&lt;p&gt;ば、馬鹿な…。glance を叩きに行ってるだと？&lt;/p&gt;

&lt;p&gt;詳しく追うの面倒なので、ざっと確認すると、S3ImageService.detail -&amp;gt; self.service.detail で、んじゃー self.service は何かしら？って言うと、S3ImageService の __init__ で以下の定義がありまして&lt;/p&gt;

&lt;pre&gt;&lt;code&gt;    def __init__(self, service=None, *args, **kwargs):
        self.cert_rpcapi = nova.cert.rpcapi.CertAPI()
        self.service = service or glance.get_default_image_service()
        self.service.__init__(*args, **kwargs)
&lt;/code&gt;&lt;/pre&gt;

&lt;p&gt;service が None のときは glance.get_default_image_service() を呼ぶから、結果 self.service には GlanceImageService が入るというわけでして、&lt;/p&gt;

&lt;p&gt;&lt;br/&gt;&lt;br/&gt;&lt;br/&gt;&lt;br/&gt;&lt;br/&gt;&lt;br/&gt;&lt;br/&gt;&lt;br/&gt;&lt;br/&gt;&lt;br/&gt;
&lt;font size=&#34;5&#34;&gt;(＃ﾟДﾟ)&lt;/font&gt;
&lt;br/&gt;&lt;br/&gt;&lt;br/&gt;&lt;br/&gt;&lt;br/&gt;&lt;br/&gt;&lt;br/&gt;&lt;br/&gt;&lt;br/&gt;&lt;br/&gt;&lt;/p&gt;

&lt;p&gt;いやもう駄目じゃん、これ、何があっても glance 呼ばれるじゃん。だって service = S3ImageService やっても循環するだけでしょ？いやいやいや、やってみたよ、無駄だとわかってもやってみたよ、修造よろしくリスペクトな気持ちでやってみたよ、そしたらね、&lt;/p&gt;

&lt;pre&gt;&lt;code&gt;2013-12-06 16:39:41.592 12145 DEBUG nova.wsgi [-] Loading app ec2 from /etc/nova/api-paste.ini load_app /usr/lib/python2.6/site-packages/nova/wsgi.py:
484
2013-12-06 16:39:41.680 12145 DEBUG nova.image.s3 [-] s3.py serivce is None __init__ /usr/lib/python2.6/site-packages/nova/image/s3.py:91
2013-12-06 16:39:41.681 12145 DEBUG nova.image.s3 [-] s3.py serivce is None __init__ /usr/lib/python2.6/site-packages/nova/image/s3.py:91
(中略)
2013-12-06 16:39:41.728 12145 DEBUG nova.image.s3 [-] s3.py serivce is None __init__ /usr/lib/python2.6/site-packages/nova/image/s3.py:91
2013-12-06 16:39:41.728 12145 DEBUG nova.image.s3 [-] s3.py serivce is None __init__ /usr/lib/python2.6/site-packages/nova/image/s3.py:91
2013-12-06 16:39:41.729 12145 CRITICAL nova [-] &#39;module&#39; object has no attribute &#39;tracebacklimit&#39;
2013-12-06 16:39:41.729 12145 TRACE nova Traceback (most recent call last):
2013-12-06 16:39:41.729 12145 TRACE nova   File &amp;quot;/usr/bin/nova-api&amp;quot;, line 10, in &amp;lt;module&amp;gt;
2013-12-06 16:39:41.729 12145 TRACE nova     sys.exit(main())
(中略)
2013-12-06 16:39:41.729 12145 TRACE nova   File &amp;quot;/usr/lib/python2.6/site-packages/nova/api/ec2/cloud.py&amp;quot;, line 226, in __init__
2013-12-06 16:39:41.729 12145 TRACE nova     self.image_service = s3.S3ImageService()
2013-12-06 16:39:41.729 12145 TRACE nova   File &amp;quot;/usr/lib/python2.6/site-packages/nova/image/s3.py&amp;quot;, line 93, in __init__
2013-12-06 16:39:41.729 12145 TRACE nova     self.service = S3ImageService()
(中略)
2013-12-06 16:39:41.729 12145 TRACE nova   File &amp;quot;/usr/lib/python2.6/site-packages/nova/image/s3.py&amp;quot;, line 91, in __init__
2013-12-06 16:39:41.729 12145 TRACE nova     LOG.debug(&amp;quot;s3.py serivce is %s&amp;quot;, service)
2013-12-06 16:39:41.729 12145 TRACE nova   File &amp;quot;/usr/lib64/python2.6/logging/__init__.py&amp;quot;, line 1305, in debug
2013-12-06 16:39:41.729 12145 TRACE nova     self.logger.debug(msg, *args, **kwargs)
2013-12-06 16:39:41.729 12145 TRACE nova   File &amp;quot;/usr/lib64/python2.6/logging/__init__.py&amp;quot;, line 1044, in debug
2013-12-06 16:39:41.729 12145 TRACE nova     self._log(DEBUG, msg, args, **kwargs)
2013-12-06 16:39:41.729 12145 TRACE nova   File &amp;quot;/usr/lib64/python2.6/logging/__init__.py&amp;quot;, line 1173, in _log
2013-12-06 16:39:41.729 12145 TRACE nova     self.handle(record)
2013-12-06 16:39:41.729 12145 TRACE nova   File &amp;quot;/usr/lib64/python2.6/logging/__init__.py&amp;quot;, line 1183, in handle
2013-12-06 16:39:41.729 12145 TRACE nova     self.callHandlers(record)
2013-12-06 16:39:41.729 12145 TRACE nova   File &amp;quot;/usr/lib64/python2.6/logging/__init__.py&amp;quot;, line 1220, in callHandlers
2013-12-06 16:39:41.729 12145 TRACE nova     hdlr.handle(record)
2013-12-06 16:39:41.729 12145 TRACE nova   File &amp;quot;/usr/lib64/python2.6/logging/__init__.py&amp;quot;, line 679, in handle
2013-12-06 16:39:41.729 12145 TRACE nova     self.emit(record)
2013-12-06 16:39:41.729 12145 TRACE nova   File &amp;quot;/usr/lib64/python2.6/logging/handlers.py&amp;quot;, line 405, in emit
2013-12-06 16:39:41.729 12145 TRACE nova     logging.FileHandler.emit(self, record)
2013-12-06 16:39:41.729 12145 TRACE nova   File &amp;quot;/usr/lib64/python2.6/logging/__init__.py&amp;quot;, line 860, in emit
2013-12-06 16:39:41.729 12145 TRACE nova     StreamHandler.emit(self, record)
2013-12-06 16:39:41.729 12145 TRACE nova   File &amp;quot;/usr/lib64/python2.6/logging/__init__.py&amp;quot;, line 804, in emit
2013-12-06 16:39:41.729 12145 TRACE nova     self.handleError(record)
2013-12-06 16:39:41.729 12145 TRACE nova   File &amp;quot;/usr/lib64/python2.6/logging/__init__.py&amp;quot;, line 733, in handleError
2013-12-06 16:39:41.729 12145 TRACE nova     traceback.print_exception(ei[0], ei[1], ei[2], None, sys.stderr)
2013-12-06 16:39:41.729 12145 TRACE nova   File &amp;quot;/usr/lib64/python2.6/traceback.py&amp;quot;, line 125, in print_exception
2013-12-06 16:39:41.729 12145 TRACE nova     print_tb(tb, limit, file)
2013-12-06 16:39:41.729 12145 TRACE nova   File &amp;quot;/usr/lib64/python2.6/traceback.py&amp;quot;, line 57, in print_tb
2013-12-06 16:39:41.729 12145 TRACE nova     if hasattr(sys, &#39;tracebacklimit&#39;):
2013-12-06 16:39:41.729 12145 TRACE nova AttributeError: &#39;module&#39; object has no attribute &#39;tracebacklimit&#39;
2013-12-06 16:39:41.729 12145 TRACE nova
&lt;/code&gt;&lt;/pre&gt;

&lt;p&gt;tracebacklimit …そりゃそうだ、だってループしてんだもん、そりゃそうだ。&lt;/p&gt;

&lt;p&gt;あーあ、終りましたよ、また今年も失敗で終りですよ。何だよ、タイトルを「Glance のかわりに Walrus を使ってみる」じゃなくて「Glance のバックエンドに Walrus を使ってみる」って書いてたら今頃きっと成功ですよ。もうね、世の OpenStacker が「ぐぐー、やっぱ Glance より Walrus かー、ちきしょー」っていう姿を夢見て走ってみたのに、気付いたら明後日の方向にスロットル全開ですよ。&lt;/p&gt;

&lt;p&gt;ちっきしょー、s3.py なんて名前でファイルを置いとくから淡い期待を抱いちゃうじゃん、もうね、s3.py じゃなくて ese3.py とかで置いておいてくれたらいいのに。&lt;/p&gt;

&lt;p&gt;&lt;br/&gt;&lt;br/&gt;&lt;font size=&#34;5&#34; color=&#34;#ee2222&#34;&gt;クリスマスまであと17日です✌(́◓q◔`)✌&lt;/font&gt;&lt;/p&gt;
&lt;div class=&#34;footnotes&#34;&gt;

&lt;hr /&gt;

&lt;ol&gt;
&lt;li id=&#34;fn:1&#34;&gt;ツールに頼らないと環境を作れないぐらいに初心者です
 &lt;a class=&#34;footnote-return&#34; href=&#34;#fnref:1&#34;&gt;&lt;sup&gt;[return]&lt;/sup&gt;&lt;/a&gt;&lt;/li&gt;
&lt;/ol&gt;
&lt;/div&gt;
</description>
    </item>
    
  </channel>
</rss>